modules/tensor_mechanics/src/materials/Compute2DIncrementalStrain.C

```diff
@@ -17,34 +17,72 @@ validParams<Compute2DIncrementalStrain>()
 {
   InputParameters params = validParams<ComputeIncrementalSmallStrain>();
   params.addClassDescription("Compute strain increment for incremental strains in 2D geometries.");
+
+  MooseEnum outOfPlaneDirection("x y z", "z");
+  params.addParam<MooseEnum>(
+      "out_of_plane_direction", outOfPlaneDirection, "The direction of the out-of-plane strain.");
   return params;
 }
 
 Compute2DIncrementalStrain::Compute2DIncrementalStrain(const InputParameters & parameters)
-  : ComputeIncrementalSmallStrain(parameters)
+  : ComputeIncrementalSmallStrain(parameters),
+    _out_of_plane_direction(getParam<MooseEnum>("out_of_plane_direction"))
+{
+}
+
+void
+Compute2DIncrementalStrain::initialSetup()
 {
+  for (unsigned int i = 0; i < 3; ++i)
+  {
+    if (_out_of_plane_direction == i)
+    {
+      _disp[i] = &_zero;
+      _grad_disp[i] = &_grad_zero;
+    }
+    else
+    {
+      _disp[i] = &coupledValue("displacements", i);
+      _grad_disp[i] = &coupledGradient("displacements", i);
+    }
+
+    if (_fe_problem.isTransient() && i != _out_of_plane_direction)
+      _grad_disp_old[i] = &coupledGradientOld("displacements", i);
+    else
+      _grad_disp_old[i] = &_grad_zero;
+  }
 }
 
 void
 Compute2DIncrementalStrain::computeTotalStrainIncrement(RankTwoTensor & total_strain_increment)
 {
   // Deformation gradient calculation for 2D problems
-  // Note: x_disp is the radial displacement, y_disp is the axial displacement
   RankTwoTensor A(
       (*_grad_disp[0])[_qp], (*_grad_disp[1])[_qp], (*_grad_disp[2])[_qp]); // Deformation gradient
   RankTwoTensor Fbar((*_grad_disp_old[0])[_qp],
                      (*_grad_disp_old[1])[_qp],
                      (*_grad_disp_old[2])[_qp]); // Old Deformation gradient
 
-  // Compute the displacement gradient (2,2) value for plane strain, generalized plane strain, or
-  // axisymmetric problems
-  A(2, 2) = computeGradDispZZ();
-  Fbar(2, 2) = computeGradDispZZOld();
+  // Compute the displacement gradient of the out of plane direction for plane strain,
+  // generalized plane strain, or axisymmetric problems
+  A(_out_of_plane_direction, _out_of_plane_direction) = computeOutOfPlaneGradDisp();
+  Fbar(_out_of_plane_direction, _out_of_plane_direction) = computeOutOfPlaneGradDispOld();
 
   _deformation_gradient[_qp] = A;
   _deformation_gradient[_qp].addIa(1.0);
 
-  A -= Fbar; // very nearly A = gradU - gradUold, adapted to cylindrical coords
+  A -= Fbar; // very nearly A = gradU - gradUold
 
   total_strain_increment = 0.5 * (A + A.transpose());
 }
+
+void
+Compute2DIncrementalStrain::displacementIntegrityCheck()
+{
+  if (_out_of_plane_direction != 2 && _ndisp != 3)
```

Comment:
It's odd that you've created a C++ enum, several MOOSE enums, and also unsigned ints to represent direction all in this PR. You might want to go for some consistency. MooseEnum has a templated method so you can convert from MooseEnum to a C++ Enum which would be ideal for all of these cases. You might look at using that.

Reply:
I was keeping it consistent with how some of the existing enums have been pasted through the TensorMechanicsAction and being converted to ints in the classes created by the action. I can look into changing these but I think there are additional places it could be used.  

I've got a patch for the valgrind issue and some of your comments regarding the tests.  Perhaps this can explored in another PR where other instances of the conversion of the enum to an int are also addressed.